core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1620,6 +1684,9 @@ class KafkaController(val config: KafkaConfig,
 
   private def processRegisterBrokerAndReelect(): Unit = {
     _brokerEpoch = zkClient.registerBroker(brokerInfo)
+    if (config.preferredController) {
+      zkClient.registerPreferredControllerId(brokerInfo.broker.id)
```

Comment:
The registration of the broker znode (BZ) and the preferred controller znode (PCZ) are not done atomically. I am concerned that this can cause potential problems because both of them are ephemeral znodes, which come and go on zk disconnect. 

In this case, for a given preferred controller broker, there are three possible combinations of these two znodes other broker can see at any given point in time:
1. Both BZ and PCZ exist
2. Both BZ and PCZ don't exist
3. BZ exists and PCZ doesn't

Case 3) can be a problem. One example: assuming a preferred controller A gets restarted and a new controller B gets elected. If 3) happens, the current controller B will think A is alive but it is not a preferred controller so new partitions can be created on preferred controller A within a short window of time.

Although we can avoid this particular race by register PCZ first and the BZ, I think it is still better to embed the preferred controller information in the broker znode (as a boolean field) and add relevant logics in `processBrokerModification` and `processBrokerChange`. What do you think?

Reply:
Discussed offline,  there are some drawbacks of embedding preferred controller information in the broker znode:
1) To ensure a proper preferred controller election and fallback controller election, a non-active broker needs to process broker change event and need to read znode data of all brokers rather only the preferred controller znode.  Broker change event occurred much more frequently than preferred controller change. 
2) we need to bump up broker znode version and inter broker protocol version in order to to prevent the following case: a) meta data cache had an information about a preferred controller (but it couldn't tell if it is preferred controller without bumping up the inter-broker protocol version). b) the preferred controller goes offline c) the broker assigned a replica to the preferred controller based on outdated meta data cache.   However, bumping up broker znode version and inter broker protocol version doesn't complete solve the problem of preferred controller will not get assigned new partitions (e.g, partition reassignment is an example. 

If we stick with current preferred controller znode approach,  we don't have the above mention issues.  In addition, in the future, we can further separated broker znode and preferred controller znode (i.e., the preferred controller only register under preferred controller znode not the broker znode). But this long term separation depends on implementing features for preferred controller node to not take any new partitions. 

Based on the above reasons, I prefer the preferred controller znode approach.

Reply:
Talked offline, the main reason to separate preferred_controller znode is to avoid having all brokers listen on the broker registration znode.